### ISSUE TITLE
Feature: support code reuse for multiple-extension repos

### ISSUE THREAD

**Issue by asbjornh** [contentful/create-contentful-extension#67]
First of all, thanks for this extremely useful tool!

We've started to use extensions a lot. One consequence of this is that our extension repo now has tons of duplicate code . On average probably something like 70% - 80% of the code for each extension is shared with the other ones (for instance we have three very similar extensions that each have something like 5 unique lines).

In a regular webpack project we'd solve this by putting common utilities and components in a folder beside the individual extension folders, and configure webpack to resolve  `node_modules` from within the respective extension folders. This does not work with `contentful-extension-scripts` – the build crashes because `node_modules` can't be resolved for files that are not beside or below `package.json` in the file tree.

For `babel` it seems sufficient to rename `.babelrc` to `babel.config.json` to enable transpilation of files outside of the root folder. I'm not familiar with Parcel, so I don't know whether this is even possible, and even if it were possible I see no documented way of passing configuration to Parcel.

Duplicating so much code for each new extension does not scale well. Having support (or a documented workaround) for code reuse would be a really great addition!

**Comment by diegoaguilar** [comment]
@asbjornh What was the solution, approach or strategy that you took

**Comment by asbjornh** [comment]
@diegoaguilar No solution really 😐 We’ve just kept duplicating code except for a couple of utility functions that don’t use `node_modules` or need transpilation.

**Comment by asbjornh** [comment]
@diegoaguilar I had another look at it today and it's actually not that hard to get around using Webpack. 

For deployment, you'd just build with Webpack, and then do the following (`contentful-cli` is required for this):

```bash
contentful extension update --force
```

One thing to be mindful of when building manually with webpack instead of using `contentful-extension-scripts` is that the plugin can only consist of a single `.html` file, so all assets (js, css etc) need to be inlined. [inline-source-webpack-plugin](https://github.com/KyLeoHC/inline-source-webpack-plugin) can be used for that.

For local development, you can use `webpack-dev-server` and then run the below command once the plugin is available on `localhost`:

```bash
contentful extension update --src http://localhost:8080 --force
```

This would make the Contentful UI load the extension from `localhost` just as when running `contentful-extension-scripts start`.

When doing this you'll probably _not_ want to inline your assets, since they'll be cached in the HTML which is not ideal for development. If you're using `inline-source-webpack-plugin` a way of doing this is to have two HTML templates (one for development and one for production), where the production one inlines assets and the development one does not.

As for importing files from "outside" of the project folder, I don't remember exactly how to do that with Webpack but I've done it in the past so I'm fairly confident that it's still possible.

**Comment by diegoaguilar** [comment]
Thanks for sharing this! So you're saying that if i use the `contentful-cli` way for building instead of using create-contentful-extension which abstracts and wraps too much, then I'll be able to set it up with Webpack.

Correct?

**Comment by asbjornh** [comment]
Well, sort of. `contentful-cli` doesn't provide a way of building stuff, but it lets you upload extensions either from a url or a file. `contentful-extension-scripts start` basically builds your code with Parcel and then calls into `contentful-cli` [here](https://github.com/contentful/create-contentful-extension/blob/44e8f650c8cf189e0aef027c5c0ba879aee18f4b/packages/contentful-extension-scripts/scripts/utils/updateExtension.js#L10).

[Here's](https://github.com/contentful/contentful-cli/tree/master/docs/extension/update) the docs for the relevant part of `contentful-cli`.

In short, if you use `contentful-cli` instead of `contentful-extension-scripts` you control the build system yourself.

After learning more about the Contentful tooling ecosystem I'm not really sure that this is a shortcoming of `contentful-extension-scripts` tbh. It would be good to mention in the documentation how to use other build systems than Parcel though.

**Comment by asbjornh** [comment]
I created a small plugin for Webpack that makes development and deployment of extensions a little easier when using Webpack: [update-contentful-extension-webpack-plugin](https://github.com/asbjornh/update-contentful-extension-webpack-plugin)

**Comment by Jwhiles** [comment]
It sounds like you've been able to find a process that works for you. 

I'd also note that we're now supporting https://www.npmjs.com/package/create-contentful-app instead of this repo. It uses create-react-app, and should be more straightforward if you need to customise the build process.

I'd also mention suggest that if you're managing a lot of apps or extensions with shared code, you might benefit from using [lerna](https://github.com/lerna/lerna) which offers some tools for managing mono-repos.

Cheers!
